File: app/src/main/java/com/newrdev/experimental/lolchampions/view/home/champions/ChampionsPresenter.java
package com.newrdev.experimental.lolchampions.view.home.champions;

import com.newrdev.experimental.lolchampions.data.ChampionRepository;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by rudolph on 5/18/17.
 */

public class ChampionsPresenter implements ChampionsContract.Presenter {
    private ChampionsContract.View view;
    private ChampionRepository repository;

    public ChampionsPresenter() {
        repository = new ChampionRepository();
    }

    @Override
    public void getAllChampions() {
        repository.getAll()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(view::showChampions, this::error);
    }

    private void error(Throwable error) {
        System.out.println("Error "+error.getLocalizedMessage());
    }

    @Override
    public void bind(ChampionsContract.View view) {
        this.view = view;
    }

    @Override
    public void unbind() {
        if (view != null) {
            view = null;
        }
    }
}

File: app/src/main/java/com/newrdev/experimental/lolchampions/view/home/champions/ChampionsPresenter.kt
package com.newrdev.experimental.lolchampions.view.home.champions

import android.util.Log.println
import com.newrdev.experimental.lolchampions.data.ChampionRepository
import com.newrdev.experimental.lolchampions.data.entity.Champion

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers

/**
 * Created by rudolph on 5/18/17.
 */

class ChampionsPresenter : ChampionsContract.Presenter {
    private var view: ChampionsContract.View? = null
    private val repository: ChampionRepository

    init {
        repository = ChampionRepository()
    }

    override fun getAllChampions() {
        repository.all
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(Consumer<List<Champion>> { view!!.showChampions(it) }, Consumer<Throwable> { this.error(it) })
    }

    private fun error(error: Throwable) {
        println(1, "Presenter", "Error " + error.localizedMessage)
    }

    override fun bind(view: ChampionsContract.View) {
        this.view = view
    }

    override fun unbind() {
        if (view != null) {
            view = null
        }
    }
}
